core/types/transaction.go

```diff
@@ -71,28 +72,35 @@ type txdataMarshaling struct {
 	S            *hexutil.Big
 }
 
-func NewTransaction(nonce uint64, to common.Address, amount *big.Int, gasLimit uint64, gasPrice *big.Int, data []byte) *Transaction {
-	return newTransaction(nonce, &to, amount, gasLimit, gasPrice, data)
+func NewTransaction(nonce uint64, to common.Address, amount *big.Int, gasLimit uint64, gasPrice *big.Int, data []byte, l1MessageSender *common.Address) *Transaction {
+	if l1MessageSender == nil {
+		return newTransaction(nonce, &to, amount, gasLimit, gasPrice, data, nil)
+	} else {
+		sender := new(common.Address)
+		*sender = *l1MessageSender
+		return newTransaction(nonce, &to, amount, gasLimit, gasPrice, data, sender)
```

Comment:
Just for my own learning, why is this logic necessary?  Seems like you want to be sure that the `sender` passed into `newTransaction` is at a different spot in memory/unique right?  Why is that needed here and not for the other. inputs like `amount`?

Reply:
Or maybe `amount` is easier as it's a built in type so I guess a better question is why not pass `&l1MessageSender` like with `&to`?

Reply:
Yeah, don't want modifications to vars that should be immutable from outside of this package. Can put the burden on all the different callers or just do it here and know for sure. If done that way everywhere else, though, I can update here to follow convention.

Reply:
I'll update it to make the caller choose. Caller can copy safely ahead of time if they think it makes sense.